### ISSUE TITLE
Options to continue consuming if _REVOKE_PARTITIONS raised

### ISSUE THREAD

**Issue by MaximGurschi** [edenhill/librdkafka#764]
Hello. 

1.In my setup i have two clusters: **Dc1** and Dc2.
2.Consumer (which is the single one in its consumer group) consumes from **Dc1**.
3.I shutdown all brokers in **Dc1**.
4.My code detects that **Dc1** is down (_ALL_BROKERS_DOWN). It stops the consumer_poll loop to **Dc1**. Then the code creates a new connection to Dc2 and consumes from there (the same group id is used to connect to Dc2). The connection to **Dc1** is kept.
5. I bring brokers from **Dc1** back up.
6.The code detects that **Dc1** is back up, stops the connection to Dc2 and switches back to **Dc1**. The consumer_poll loop for **Dc1** is resumed.
7._REVOKE_PARTITIONS gets raised for the connection to **Dc1**.

How can i force the consumer to continue consuming from **Dc1** after point 7?

**Comment by MaximGurschi** [comment]
What i sometimes see is that _ASSIGN_PARTITIONS sometimes happens immediately after _REVOKE_PARTITIONS. This is the ideal case as i get the intended results. The consumer resumes getting messages from **Dc1**.

However that is not always the case and once the _REVOKE_PARTITIONS happens then there looks like there is nothing i can do to start consuming. Is the only resort to create a new connection to **Dc1** and not reuse the previous one?

_On a side note I also noticed that the _ALL_BROKERS_DOWN does not happen for an indefinite time even if all brokers are shut down. Sometimes i have to bring them back up and down again to trigger this event._

**Comment by MaximGurschi** [comment]
Also will calling **rd_kafka_destroy** indeed block indefinitely trying to commit offsets? If my entire cluster is down for unbounded time, will this call block for as long?

**Comment by MaximGurschi** [comment]
It is this condition that is never satisfied:

#define rd_kafka_broker_terminating(rkb) \
        (rd_refcnt_get(&(rkb)->rkb_refcnt) <= 1)

The call is from "static int rd_kafka_broker_thread_main (void *arg)". For some reason the ref count is above 1 when the cluster brokers are down.

**Comment by MaximGurschi** [comment]
I see a similar behaviour for the producer. If I destroy the producer while none of the cluster brokers are available, the call to destroy hangs.

**Comment by MaximGurschi** [comment]
Hello Eden. Just to summarize for you the two questions:

1.If the consumer gets a _REVOKE_PARTITIONS event, to force the consumer to continue consuming, do i have to create a new consumer handle?
2.Will destroying consumer/producer block while there are pending commits/sends? If the whole cluster is down then this will block for as long?

**Comment by edenhill** [comment]
Sorry for tending to this earlier, thanks for the summary.

1. Multiple ASSIGN + REVOKE sequences are natural in the lifetime of a consumer. As soon as the group rebalances (typically because a member joins or leaves) all members will REVOKE their current assignment and await a new ASSIGNMENT. So to answer your question: no, you should hold on to your existing consumer and let it rebalance (which can take up to session.timeout.ms).

2. Yes, rd_kafka_consumer_close() will block for up to ~session.timeout.ms, see here:
https://github.com/edenhill/librdkafka/blob/master/src/rdkafka.h#L2032

**Comment by MaximGurschi** [comment]
Oh, great, thx for replying.

1.Tomorrow i will try to wait longer for the consumer to receive a **_ASSIGN_PARTITIONS** after **_REVOKE_PARTITIONS**. Will update on this.
2.Yes i looked through the source code and saw the comment about **session.timeout.ms**, I had it configured for 30 seconds. But in practice i have a producer that is blocked for 3 hours now on this loop that i run before destroy handle:
_while (rd_kafka_outq_len(rk) > 0)  rd_kafka_poll(rk, 50);_
Is **this** loop supposed to block indefinitely? If i bring the cluster back up just for this loop to complete then the following handle destroy call takes about 10 minutes (regardless if the cluster is up or down at the destroy step).

**Comment by edenhill** [comment]
The producer should not block on rd_kafka_destroy() even if there are outstanding messages waiting to be delivered, but you will loose those messages.
What is the full config you provide to the producer?

**Comment by MaximGurschi** [comment]
These are the two settings i have now for the producer (topic) that is blocked on the **poll loop** for 3h+ (still running):
compression.codec=snappy
message.timeout.ms=5000

**rd_kafka_outq_len** gives me 3 each time. Will this loop never complete while the cluster is down? Previous testing has shown that if i bring it back up then the loop completes but the next destroy takes ~10 minutes.

**Comment by ah-** [comment]
If it's just waiting for the timeout to expire how about some method that will cancel all outstanding requests?

**Comment by MaximGurschi** [comment]
Ok, came back today and the loop is still running. That is 15 hours now. **rd_kafka_outq_len** is still 3.

The version of librdkafka is 0.9.1.

**Comment by edenhill** [comment]
Can you extract a core file from that process? (using gcore or gdb)

**Comment by MaximGurschi** [comment]
Erm, not any more. I closed it :/ See my updated comment above. Would a code sample work?

**Comment by edenhill** [comment]
Absolutely!

**Comment by MaximGurschi** [comment]
Sorry for the delay. Needed to finish up something. Please have a look at this. I am repeating myself but do you expect the clean up while loop to complete?

The below code, when using current master of librdkafka produces a hang on the:
while (rd_kafka_outq_len(rk) > 0) {
		rd_kafka_poll(rk, 100);
		printf("polling\n");
	}

, loop if the cluster was shutdown while we were still producing.

`int main()
{
	//0.Create config.
	char errstr[512];
	rd_kafka_conf_t *conf = rd_kafka_conf_new();
	rd_kafka_conf_set(conf, "compression.codec", "snappy", errstr, sizeof(errstr)); printf("%% : %s\n", errstr);
	
	//1.Create producer.
	rd_kafka_t *rk = rd_kafka_new(RD_KAFKA_PRODUCER, conf, errstr, sizeof(errstr));
	if(!rk) printf("%% Failed to create new producer: %s\n", errstr);
	char *brokers = "**YOUR_CLUSTER**";
	if (rd_kafka_brokers_add(rk, brokers) == 0) fprintf(stderr, "%% No valid brokers specified\n");

	//2.Create topic.
	rd_kafka_topic_t *rkt;
	rd_kafka_topic_conf_t *topic_conf= rd_kafka_topic_conf_new();
	rd_kafka_topic_conf_set(topic_conf, "message.timeout.ms", "5000", errstr, sizeof(errstr)); printf("%% : %s\n", errstr);
	rkt = rd_kafka_topic_new(rk, "ABC", topic_conf);

	//3.Produce some messages.
	**//The idea is to shutdown the cluster while we are still producing.**
	int shouldRun = 50;
	while (--shouldRun >= 0)
	{
		int errorCode = rd_kafka_produce(rkt, 0, RD_KAFKA_MSG_F_COPY, NULL, 0, NULL, 0, NULL);
		if (errorCode == -1)
			printf("%% Failed to produce to topic %s partition %i: %s\n", rd_kafka_topic_name(rkt), 0, rd_kafka_err2str(rd_kafka_last_error()));
		else
			printf("produce is good\n");
		std::this_thread::sleep_for(std::chrono::milliseconds(500));
	}
	printf("no more sends\n");

	//4.Cleanup.
	**//If the cluster was shutdown then my tests show that this loop never completes.**
	struct timeb start, end;
	ftime(&start);
	**while (rd_kafka_outq_len(rk) > 0) {
		rd_kafka_poll(rk, 100);
		printf("polling\n");
	}**
	ftime(&end);
	int diff = (int)(1000.0 * (end.time - start.time) + (end.millitm - start.millitm));
	**printf("\nwhile %u milliseconds\n", diff);**

	printf("rd_kafka_topic_destroy\n");
	ftime(&start);
	**rd_kafka_topic_destroy(rkt);**
	ftime(&end);
	diff = (int)(1000.0 * (end.time - start.time) + (end.millitm - start.millitm));
	**printf("\nrd_kafka_topic_destroy %u milliseconds\n", diff);**
	
	ftime(&start);
	**rd_kafka_destroy(rk);**
	ftime(&end);
	diff = (int)(1000.0 * (end.time - start.time) + (end.millitm - start.millitm));
	**printf("\nrd_kafka_destroy %u milliseconds\n", diff);**

	printf("done\n");
    return 0;
}`

**Comment by MaximGurschi** [comment]
I have the above running for 40 minutes now still the loop does not stop. The **rd_kafka_curr_msgs_cnt** portion keeps returning non zero.

**Comment by MaximGurschi** [comment]
Also I have experimented to see if **_ASSIGN_PARTITIONS** gets triggered after **_REVOKE_PARTITIONS**. 

It is now taking 7 minutes and the **_ASSIGN_PARTITIONS** does not happen. Is this expected?

**Comment by MaximGurschi** [comment]
In case there is any confusion, I am running on Windows.

**Comment by MaximGurschi** [comment]
To make it easier for you Eden i will factor out the other question about rebalance in a separate issue.

Thus for this issue, the only question is around the fact that:
while (rd_kafka_outq_len(rk) > 0) {
rd_kafka_poll(rk, 100);
printf("polling\n");
}

, never terminates if the cluster is down.

**Comment by MaximGurschi** [comment]
While i have your attention, predictably, i am asking you about this one :)

Can you please provide some feedback?

Should the above poll loop terminate even if the whole cluster is down?

**Comment by edenhill** [comment]
I think it should terminate within 5s (message.timeout.ms).
Will try to reproduce it.

**Comment by MaximGurschi** [comment]
Thank you! Please try.

The **rd_kafka_curr_msgs_cnt** returning above zero never terminates the loop for me.

**Comment by MaximGurschi** [comment]
Did you manage to reproduce this?

**Comment by edenhill** [comment]
I fixed a similar issue a couple of days ago, where it wouldnt time out messages waiting in the broker retransmit queue.
Can you update to latest master and give it a spin?

**Comment by MaximGurschi** [comment]
Looks like it works! Thank you. Can you point me at the commit?

Also can Andreas push this out through RdKafka or is this not a stable release yet?

**Comment by edenhill** [comment]
https://github.com/edenhill/librdkafka/commit/52ef556dd3d655ebff1bbf83a49732e95b63f9a5

There's a bunch of fixes and enhancements added recently so I wouldnt consider master as ready for production yet.

@ah-
